### PR TITLE
Add Timeouts to Tracker Client

### LINKED ISSUE #676: Tracker Checker:  Improve errors. More concrete errors
Parent issue: https://github.com/torrust/torrust-tracker/issues/669
Relates to:
- https://github.com/torrust/torrust-tracker/issues/682
- https://github.com/torrust/torrust-tracker/issues/679
- https://github.com/torrust/torrust-tracker/issues/678
- https://github.com/torrust/torrust-tracker/issues/683

Currently, you can run the Tracker Checker with this input:

```console
TORRUST_CHECKER_CONFIG='{
    "udp_trackers": ["144.126.245.19:6969"],
    "http_trackers": ["https://tracker.torrust-demo.com"],
    "health_checks": ["https://tracker.torrust-demo.com/health_check"]
}' cargo run --bin tracker_checker
```

When something is wrong you only get three types of errors or the application panics.

For example:

```console
TORRUST_CHECKER_CONFIG='{
    "udp_trackers": [],
    "http_trackers": [],
    "health_checks": ["https://localhost:1515"]
}' cargo run --bin tracker_checker
```

Output:

```
Running checks for trackers ...
UDP trackers ...
HTTP trackers ...
Health checks ...
✗ - Health API at https://localhost:1515/ is failing: reqwest::Error { kind: Request, url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(1515), path: "/", query: None, fragment: None }, source: hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })) }
```

Errors are defined with this enum:

```rust
#[derive(Debug)]
pub enum CheckError {
    UdpError { socket_addr: SocketAddr },
    HttpError { url: Url },
    HealthCheckError { url: Url },
}
```

But depending on the service there could be different reasons. For example:

For the Health Check:

- The connection can be refused.
- The response status code could not be 200.
- Etcetera.

For the UDP tracker:

- You could not bind the client to a local port.
- You could not send a packet to the remote server.
- Etcetera.

We should introduce a second level of types for errors. For example:

```rust
#[derive(Debug)]
pub enum CheckError {
    Udp { error: UdpError },
    Http { error: HttpError },
    HealthCheck { error: HealthCheckError },
}

#[derive(Debug)]
pub enum HealthCheckError {
    ConnectionRefused { url: Url },
    NotOkResponse { url: Url, status: StatusCode },
    HealthCheck { error: HealthCheckError },
}

// ...
```

This issue should be implemented after implementing other issues related to timeout errors. Because those issues might introduce new errors.

### LINKED ISSUE #677: Tracker Checker: gracefully handle timeouts
Parent issue: https://github.com/torrust/torrust-tracker/issues/669

Currently, you can run the Tracker Checker with this input:

```console
TORRUST_CHECKER_CONFIG='{
    "udp_trackers": ["127.0.0.1:6969"],
    "http_trackers": ["http://127.0.0.1:7070"],
    "health_checks": ["http://127.0.0.1:1313/health_check"]
}' cargo run --bin tracker_checker
```

In some cases, there could be a missing timeout for network operations and in other cases where there is a timeout, the error is not handled gracefully.

This issue tracks all subtasks related to timeout issues.

### Subtasks

- [x] https://github.com/torrust/torrust-tracker/issues/678
- [x] https://github.com/torrust/torrust-tracker/issues/679
- [x] https://github.com/torrust/torrust-tracker/issues/682

### LINKED ISSUE #678: Tracker Checker: handle Health Check timeouts
Parent issue: https://github.com/torrust/torrust-tracker/issues/677

You can run a Tracker Checker with:

```console
TORRUST_CHECKER_CONFIG='{
    "udp_trackers": [],
    "http_trackers": [],
    "health_checks": ["http://127.0.0.1:3030/sleep"]
}' cargo run --bin tracker_checker
```

If the server does not reply in 5 seconds you will see an error like this:

```console
$ TORRUST_CHECKER_CONFIG='{
    "udp_trackers": [],
    "http_trackers": [],
    "health_checks": ["http://127.0.0.1:3030/sleep"]
}' cargo run --bin tracker_checker
    Finished dev [optimized + debuginfo] target(s) in 0.08s
     Running `target/debug/tracker_checker`
Running checks for trackers ...
UDP trackers ...
HTTP trackers ...
Health checks ...
✗ - Health API at http://127.0.0.1:3030/sleep is failing: reqwest::Error { kind: Request, url: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(3030), path: "/sleep", query: None, fragment: None }, source: TimedOut }
```

You can use this [example app](https://github.com/josecelano/rust-sleep) to simulate the server.

This is the function for the check:

```rust
async fn run_health_check(url: Url, console: &Console) -> Result<(), CheckError> {
    let client = HttpClient::builder().timeout(Duration::from_secs(5)).build().unwrap();

    let colored_url = url.to_string().yellow();

    match client.get(url.clone()).send().await {
        Ok(response) => {
            if response.status().is_success() {
                console.println(&format!("{} - Health API at {} is OK", "✓".green(), colored_url));
                Ok(())
            } else {
                console.eprintln(&format!(
                    "{} - Health API at {} is failing: {:?}",
                    "✗".red(),
                    colored_url,
                    response
                ));
                Err(CheckError::HealthCheckError { url })
            }
        }
        Err(err) => {
            console.eprintln(&format!(
                "{} - Health API at {} is failing: {:?}",
                "✗".red(),
                colored_url,
                err
            ));
            Err(CheckError::HealthCheckError { url })
        }
    }
}
```

It contains this line:

```rust
let client = HttpClient::builder().timeout(Duration::from_secs(5)).build().unwrap();
```

We have to catch the error instead of unwrapping. Otherwise, the Tracker Checker is halted. The Tracker Checker should finish the whole report even if one service is not available.

### LINKED ISSUE #679: Tracker Checker: handle HTTP Tracker timeouts
Parent issue: https://github.com/torrust/torrust-tracker/issues/677

You can run a Tracker Checker with:

```console
TORRUST_CHECKER_CONFIG='{
    "udp_trackers": [],
    "http_trackers": ["http://127.0.0.1:7070"],
    "health_checks": []
}' cargo run --bin tracker_checker
```

The HTTP Tracker client does not have any timeout so it will wait forever.

We have to add a `timeout` to the HTTP client requests:

```rust
/// HTTP Tracker Client
pub struct Client {
    base_url: Url,
    reqwest: ReqwestClient,
    key: Option<Key>,
}

impl Client {
    // ...

    pub async fn announce(&self, query: &announce::Query) -> Response {
        self.get(&self.build_announce_path_and_query(query)).await
    }

    pub async fn scrape(&self, query: &scrape::Query) -> Response {
        self.get(&self.build_scrape_path_and_query(query)).await
    }

    pub async fn announce_with_header(&self, query: &Query, key: &str, value: &str) -> Response {
        self.get_with_header(&self.build_announce_path_and_query(query), key, value)
            .await
    }

    pub async fn health_check(&self) -> Response {
        self.get(&self.build_path("health_check")).await
    }

    /// # Panics
    ///
    /// This method fails if there was an error while sending request.
    pub async fn get(&self, path: &str) -> Response {
        self.reqwest.get(self.build_url(path)).send().await.unwrap()
    }

    /// # Panics
    ///
    /// This method fails if there was an error while sending request.
    pub async fn get_with_header(&self, path: &str, key: &str, value: &str) -> Response {
        self.reqwest
            .get(self.build_url(path))
            .header(key, value)
            .send()
            .await
            .unwrap()
    }

    // ...
}
```

You can add a timeout to all `reqwest.get` like this in the Health Check:

```rust
HttpClient::builder().timeout(Duration::from_secs(5)).build().unwrap();
```

You can force the error by adding a `sleep` in the handler:

```rust
async fn handle(
    tracker: &Arc<Tracker>,
    announce_request: &Announce,
    client_ip_sources: &ClientIpSources,
    maybe_key: Option<Key>,
) -> Response {
    thread::sleep(time::Duration::from_secs(7));

    let announce_data = match handle_announce(tracker, announce_request, client_ip_sources, maybe_key).await {
        Ok(announce_data) => announce_data,
        Err(error) => return error.into_response(),
    };
    build_response(announce_request, announce_data)
}
```

This issue should be implemented after refactoring the HTTP Tracker client to return errors instead of panicking.

### LINKED ISSUE #680: Refactor: `bit_torrent::tracker::http::client::Client` to return errors instead of panicking
The HTTP Tracker Client struct contains some `unwrap` calls.

```rust
pub struct Client {
    base_url: Url,
    reqwest: ReqwestClient,
    key: Option<Key>,
}

impl Client {
    // ...

    pub fn bind(base_url: Url, local_address: IpAddr) -> Self {
        Self {
            base_url,
            reqwest: reqwest::Client::builder().local_address(local_address).build().unwrap(),
            key: None,
        }
    }

    // ...

    pub fn authenticated(base_url: Url, key: Key) -> Self {
        Self {
            base_url,
            reqwest: reqwest::Client::builder().build().unwrap(),
            key: Some(key),
        }
    }

    // ...

    pub async fn get(&self, path: &str) -> Response {
        self.reqwest.get(self.build_url(path)).send().await.unwrap()
    }

    // ...

    pub async fn get_with_header(&self, path: &str, key: &str, value: &str) -> Response {
        self.reqwest
            .get(self.build_url(path))
            .header(key, value)
            .send()
            .await
            .unwrap()
    }

    // ...
}
```

That makes it impossible to catch those errors upstream. For example, the Tracker Checker can not use this `Client` and continue its execution if an HTTP Tracker request fails. See, for example, this issue: https://github.com/torrust/torrust-tracker/issues/678.

We should return the `Result` type and let the caller handle the problem.

This Client was introduced for testing purposes and later moved to production, so it lacks some important features like:

- Testing.
- Better error handling.

### LINKED ISSUE #681: Refactor `bit_torrent::tracker::udp::client::UdpClient` to return errors instead of panicking
The UDP Tracker Client struct contains some `unwrap` and `panic` calls.

```rust
pub struct UdpClient {
    /// The socket to connect to
    pub socket: Arc<UdpSocket>,

    /// Timeout for sending and receiving packets
    pub timeout: Duration,
}

impl UdpClient {
    pub async fn bind(local_address: &str) -> Self {
        let valid_socket_addr = local_address
            .parse::<SocketAddr>()
            .unwrap_or_else(|_| panic!("{local_address} is not a valid socket address"));

        let socket = UdpSocket::bind(valid_socket_addr).await.unwrap();

        Self {
            socket: Arc::new(socket),
            timeout: DEFAULT_TIMEOUT,
        }
    }

    pub async fn connect(&self, remote_address: &str) {
        let valid_socket_addr = remote_address
            .parse::<SocketAddr>()
            .unwrap_or_else(|_| panic!("{remote_address} is not a valid socket address"));

        match self.socket.connect(valid_socket_addr).await {
            Ok(()) => debug!("Connected successfully"),
            Err(e) => panic!("Failed to connect: {e:?}"),
        }
    }

    pub async fn send(&self, bytes: &[u8]) -> usize {
        debug!(target: "UDP client", "sending {bytes:?} ...");

        match time::timeout(self.timeout, self.socket.writable()).await {
            Ok(writable_result) => match writable_result {
                Ok(()) => (),
                Err(e) => panic!("{}", format!("IO error waiting for the socket to become readable: {e:?}")),
            },
            Err(e) => panic!("{}", format!("Timeout waiting for the socket to become readable: {e:?}")),
        };

        match time::timeout(self.timeout, self.socket.send(bytes)).await {
            Ok(send_result) => match send_result {
                Ok(size) => size,
                Err(e) => panic!("{}", format!("IO error during send: {e:?}")),
            },
            Err(e) => panic!("{}", format!("Send operation timed out: {e:?}")),
        }
    }

    pub async fn receive(&self, bytes: &mut [u8]) -> usize {
        debug!(target: "UDP client", "receiving ...");

        match time::timeout(self.timeout, self.socket.readable()).await {
            Ok(readable_result) => match readable_result {
                Ok(()) => (),
                Err(e) => panic!("{}", format!("IO error waiting for the socket to become readable: {e:?}")),
            },
            Err(e) => panic!("{}", format!("Timeout waiting for the socket to become readable: {e:?}")),
        };

        let size = match time::timeout(self.timeout, self.socket.recv(bytes)).await {
            Ok(recv_result) => match recv_result {
                Ok(size) => size,
                Err(e) => panic!("{}", format!("IO error during send: {e:?}")),
            },
            Err(e) => panic!("{}", format!("Receive operation timed out: {e:?}")),
        };

        debug!(target: "UDP client", "{size} bytes received {bytes:?}");

        size
    }
}
```

That makes it impossible to catch those errors upstream. For example, the Tracker Checker can not use this `UdpClient` and continue its execution if a UDP Tracker request fails.

We should return the `Result` type and let the caller handle the problem.

This Client was introduced for testing purposes and later moved to production, so it lacks some important features like:

- Testing.
- Better error handling.

### LINKED ISSUE #682: Tracker Checker: handle UDP Tracker timeouts
Parent issue: https://github.com/torrust/torrust-tracker/issues/677
Depends on: https://github.com/torrust/torrust-tracker/issues/681

You can run a Tracker Checker with:

```console
TORRUST_CHECKER_CONFIG='{
    "udp_trackers": ["127.0.0.1:6969"],
    "http_trackers": [],
    "health_checks": []
}' cargo run --bin tracker_checker
```

The UDP Tracker client does have a timeout for sending and receiving responses. However, it panics when times out. See https://github.com/torrust/torrust-tracker/issues/681.

After implementing the issue https://github.com/torrust/torrust-tracker/issues/681 we should bubble the error up to this point:

(`torrust_tracker::console::clients::udp::checker::Client`)

```rust
pub struct Client {
    local_binding_address: Option<SocketAddr>,
    local_bound_address: Option<SocketAddr>,
    remote_socket: Option<SocketAddr>,
    udp_tracker_client: Option<UdpTrackerClient>,
}

impl Client {
    pub async fn bind_and_connect(&mut self, local_port: u16, remote_socket_addr: &SocketAddr) -> anyhow::Result<SocketAddr> {
        let bound_to = self.bind(local_port).await?;
        self.connect(remote_socket_addr).await?;
        Ok(bound_to)
    }

    async fn bind(&mut self, local_port: u16) -> anyhow::Result<SocketAddr> {
        let local_bind_to = format!("0.0.0.0:{local_port}");
        let binding_address = local_bind_to.parse().context("binding local address")?;

        debug!("Binding to: {local_bind_to}");
        let udp_client = UdpClient::bind(&local_bind_to).await;

        let bound_to = udp_client.socket.local_addr().context("bound local address")?;
        debug!("Bound to: {bound_to}");

        self.local_binding_address = Some(binding_address);
        self.local_bound_address = Some(bound_to);

        self.udp_tracker_client = Some(UdpTrackerClient { udp_client });

        Ok(bound_to)
    }

    async fn connect(&mut self, tracker_socket_addr: &SocketAddr) -> anyhow::Result<()> {
        debug!("Connecting to tracker: udp://{tracker_socket_addr}");

        match &self.udp_tracker_client {
            Some(client) => {
                client.udp_client.connect(&tracker_socket_addr.to_string()).await;
                self.remote_socket = Some(*tracker_socket_addr);
                Ok(())
            }
            None => Err(ClientError::NotBound.into()),
        }
    }

    pub async fn send_connection_request(&self, transaction_id: TransactionId) -> anyhow::Result<ConnectionId> {
        debug!("Sending connection request with transaction id: {transaction_id:#?}");

        let connect_request = ConnectRequest { transaction_id };

        match &self.udp_tracker_client {
            Some(client) => {
                client.send(connect_request.into()).await;

                let response = client.receive().await;

                debug!("connection request response:\n{response:#?}");

                match response {
                    Response::Connect(connect_response) => Ok(connect_response.connection_id),
                    _ => Err(ClientError::UnexpectedConnectionResponse.into()),
                }
            }
            None => Err(ClientError::NotConnected.into()),
        }
    }

    pub async fn send_announce_request(
        &self,
        connection_id: ConnectionId,
        transaction_id: TransactionId,
        info_hash: TorrustInfoHash,
        client_port: Port,
    ) -> anyhow::Result<Response> {
        debug!("Sending announce request with transaction id: {transaction_id:#?}");

        let announce_request = AnnounceRequest {
            connection_id,
            transaction_id,
            info_hash: InfoHash(info_hash.bytes()),
            peer_id: PeerId(*b"-qB00000000000000001"),
            bytes_downloaded: NumberOfBytes(0i64),
            bytes_uploaded: NumberOfBytes(0i64),
            bytes_left: NumberOfBytes(0i64),
            event: AnnounceEvent::Started,
            ip_address: Some(Ipv4Addr::new(0, 0, 0, 0)),
            key: PeerKey(0u32),
            peers_wanted: NumberOfPeers(1i32),
            port: client_port,
        };

        match &self.udp_tracker_client {
            Some(client) => {
                client.send(announce_request.into()).await;

                let response = client.receive().await;

                debug!("announce request response:\n{response:#?}");

                Ok(response)
            }
            None => Err(ClientError::NotConnected.into()),
        }
    }

    pub async fn send_scrape_request(
        &self,
        connection_id: ConnectionId,
        transaction_id: TransactionId,
        info_hashes: Vec<TorrustInfoHash>,
    ) -> anyhow::Result<Response> {
        debug!("Sending scrape request with transaction id: {transaction_id:#?}");

        let scrape_request = ScrapeRequest {
            connection_id,
            transaction_id,
            info_hashes: info_hashes
                .iter()
                .map(|torrust_info_hash| InfoHash(torrust_info_hash.bytes()))
                .collect(),
        };

        match &self.udp_tracker_client {
            Some(client) => {
                client.send(scrape_request.into()).await;

                let response = client.receive().await;

                debug!("scrape request response:\n{response:#?}");

                Ok(response)
            }
            None => Err(ClientError::NotConnected.into()),
        }
    }
}
```

We have to catch the error and add new types of errors:

```rust
#[derive(Error, Debug)]
pub enum ClientError {
    #[error("Local socket address is not bound yet. Try binding before connecting.")]
    NotBound,
    #[error("Not connected to remote tracker UDP socket. Try connecting before making requests.")]
    NotConnected,
    #[error("Unexpected response while connecting the the remote server.")]
    UnexpectedConnectionResponse,
}
```

At the top level the Tracker Checker should return a check error in the final report instead of panicking when a UDP client panics.

### LINKED ISSUE #683: Tracker Checker: handle HTTP Tracker error `ConnectionRefused`
Parent issue: https://github.com/torrust/torrust-tracker/issues/669
Depends on: https://github.com/torrust/torrust-tracker/issues/680 

When you run the Tracker Checker with a HTTP Tracker service not running you get this error:

```console
TORRUST_CHECKER_CONFIG='{
    "udp_trackers": [],
    "http_trackers": ["http://127.0.0.1:7070"],
    "health_checks": []
}' cargo run --bin tracker_checker
```

```console
Running checks for trackers ...
UDP trackers ...
HTTP trackers ...
thread 'main' panicked at /home/josecelano/Documents/git/committer/me/github/torrust/torrust-tracker/src/shared/bit_torrent/tracker/http/client/mod.rs:87:61:
called `Result::unwrap()` on an `Err` value: reqwest::Error { kind: Request, url: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(7070), path: "/announce", query: Some("info_hash=%9C8B%22%13%E3%0B%FF%21%2B0%C3%60%D2o%9A%02%13d%22&peer_addr=192.168.1.88&downloaded=0&uploaded=0&peer_id=%2DqB00000000000000001&port=17548&left=0&event=completed&compact=0"), fragment: None }, source: hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })) }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

After implementing the issue https://github.com/torrust/torrust-tracker/issues/680 we could catch the error and continue with the execution of the Tracker Checker, checking other services. CUrrently, the checker is halted without finishing.

### LINKED ISSUE #767: De-duplicate Integration Testing and Tracker Client Code
part of:
- #669

 `src/shared/bit_torrent/tracker/http/client` is mostly a copy-paste of the infrastructure code for: `tests/servers/http/`

De-duplicating this code would mean that our client code would be tested with our integration tests.

### PR DESCRIPTION
**Pull request by da2ce7**
closes: #677 closes: #682 closes: #681 closes: #679 closes: #680 closes: #678 closes: #683 closes: #676 closes: #767

part of epic:
- #669

todo:
- [ ] Write More Tests
- [x] Do Code Review
- [ ] Make Timeouts Configurable for the UDP client, like the Http Client.

done:
 - [x] https://github.com/torrust/torrust-tracker/issues/677
 - [x] https://github.com/torrust/torrust-tracker/issues/682
 - [x] https://github.com/torrust/torrust-tracker/issues/767
 - [x] https://github.com/torrust/torrust-tracker/issues/681
 - [x] https://github.com/torrust/torrust-tracker/issues/679
 - [x] https://github.com/torrust/torrust-tracker/issues/680
 - [x] https://github.com/torrust/torrust-tracker/issues/678
 - [x] https://github.com/torrust/torrust-tracker/issues/683
 - [x] https://github.com/torrust/torrust-tracker/issues/676

### CODE REVIEW

**Comment by da2ce7** [comment]
@josecelano This Pull Request is Ready for Review, however it needs some polish before merging. :)

File: src/console/clients/checker/checks/udp.rs [hunk 1]
```diff
@@ -9,79 +9,121 @@ use torrust_tracker_primitives::info_hash::InfoHash;
...
-                "✗".red(),
-                colored_tracker_url
-            ));
+        // Do Scrape
+        if {
+            let res = check_udp_scrape(&client, &ctx, vec![info_hash]).await;
+
+            check_results.push(match res {
+                Ok(_) => {
+                    console.println(&format!("{} - Announce of {} is OK", "✓".green(), colored_addr));
```
Comment (da2ce7) [inline src/console/clients/checker/checks/udp.rs]: Note: Should Be "Scrape" in Console Message Here

File: packages/configuration/src/lib.rs [hunk 2]
```diff
@@ -246,6 +247,19 @@ use torrust_tracker_primitives::{DatabaseDriver, TrackerMode};
...
+
+/// The maximum number of bytes in a UDP packet.
+pub const MAX_PACKET_SIZE: usize = 1496;
+
+/// The a free port is dynamically chosen by the operating system.
+pub const PORT_ASSIGNED_BY_OS: u16 = 0;
+
+/// A magic 64-bit integer constant defined in the protocol that is used to
+/// identify the protocol.
+pub const PROTOCOL_ID: i64 = 0x0417_2710_1980;
```
Comment (josecelano) [inline packages/configuration/src/lib.rs]: hi @da2ce7 in this context I would rename the const to `UDP_PROTOCOL_ID` or move the const to a UDP mod.

File: packages/configuration/src/lib.rs
Comment (josecelano) [inline packages/configuration/src/lib.rs:1]: Hi @da2ce7 I think this mod is too big. I would create a mod for each service and only put things in the root mod that are common to all services.

File: src/console/clients/http/app.rs [hunk 3]
```diff
@@ -14,22 +14,21 @@
 //! cargo run --bin http_tracker_client scrape http://127.0.0.1:7070 9c38422213e30bff212b30c360d26f9a02136422 | jq
```
Comment (josecelano) [inline src/console/clients/http/app.rs:14]: @da2ce7 update this to:

```
cargo run --bin http_tracker_client 10 scrape http://127.0.0.1:7070 9c38422213e30bff212b30c360d26f9a02136422 | jq
```

Where `10` is the timeout.

Comment (josecelano) [reply src/console/clients/http/app.rs:14]: Maybe we can use this as an example:

```console
cargo run --bin http_tracker_client -- --help
```

That never changes.

**Review by josecelano — DISMISSED** [review]
Hi @da2ce7 overall I think this is very good refactor. I'm going to merge it. However, we should discuss future related work.

See my comments [here](https://github.com/torrust/torrust-tracker/issues/767).

I think:

- We should move the mod `src/shared/bit_torrent` to a new package (`packages/bit_torrent`).
- That `bit_torrent` package could be splitted into several ones. See https://github.com/torrust/torrust-tracker/issues/753.
- Maybe the server and client of the same server should go in the same package. I did it on the Index APIhttps://github.com/torrust/torrust-index/tree/develop/src/web/api. Changes in the server API imply changes in the client. Besides we are versioning the API, I think it would be easier to maintain the client if it's closer to the server code. Maybe the server and client could be features of the same package otherwise you would need three packages per service: primitives, server and client. If we use only one package it could by default expose primitives and the client and the server could be features disabled by default. What do you think? Having an independent package for each service would allow other users to use them and even run a single service if they want. FOr example, some users might want to run only the UDP tracker.
- For the HTTP client requests and responses are still duplicated. If we move the server and client to the same package we can remove that duplication. They are duplicated because I did not want to couple the client to the server implementation, but when both the server and the client depend on the HTTP tracker primitives that would not be a problem. Althougth it could be the case we want them to be duplicated because there are some aspects of the types in the server we don't want to expose in the API.

**Comment by josecelano** [comment]
ACK 2a72445f1004f4d417ed1b41ebec15881cdfa4d4

**Comment by da2ce7** [comment]
needs impossible rebase, closed by #923

**Pull request closed without merging**